src/parcsr_ls/par_mgr_interp.c

```diff
@@ -2375,3 +2375,139 @@ hypre_MGRTruncateAcfCPR(hypre_ParCSRMatrix  *A_CF,
+ * hypre_MGRBuildRFromW
+ *
+ * Constructs a classical restriction operator as R = [-W I].
+ *--------------------------------------------------------------------------*/
+
+HYPRE_Int
+hypre_MGRBuildRFromW(HYPRE_Int            *C_map,
+                     HYPRE_Int            *F_map,
+                     HYPRE_BigInt         *row_starts_R,
+                     HYPRE_BigInt         *col_starts_R,
+                     hypre_ParCSRMatrix   *W,
+                     hypre_ParCSRMatrix  **R_ptr)
+{
+   /* Input matrix variables */
+   MPI_Comm               comm              = hypre_ParCSRMatrixComm(W);
+   HYPRE_BigInt           num_rows_W        = hypre_ParCSRMatrixGlobalNumRows(W);
+   HYPRE_BigInt           num_cols_W        = hypre_ParCSRMatrixGlobalNumCols(W);
+   HYPRE_MemoryLocation   memory_location_W = hypre_ParCSRMatrixMemoryLocation(W);
+
+   hypre_CSRMatrix       *W_diag            = hypre_ParCSRMatrixDiag(W);
+   HYPRE_Int              W_diag_num_rows   = hypre_CSRMatrixNumRows(W_diag);
+   HYPRE_Int              W_diag_nnz        = hypre_CSRMatrixNumNonzeros(W_diag);
+
+   hypre_CSRMatrix       *W_offd            = hypre_ParCSRMatrixOffd(W);
+   HYPRE_Int              W_offd_num_cols   = hypre_CSRMatrixNumCols(W_offd);
+   HYPRE_Int              W_offd_nnz        = hypre_CSRMatrixNumNonzeros(W_offd);
+
+   /* Output matrix */
+   hypre_ParCSRMatrix    *R;
+   HYPRE_BigInt           num_rows_R        = num_rows_W;
+   HYPRE_BigInt           num_cols_R        = num_cols_W + num_rows_W;
+   HYPRE_Int              num_nonzeros_diag = W_diag_nnz + W_diag_num_rows;
+   HYPRE_Int              num_nonzeros_offd = W_offd_nnz;
+
+   /* Sanity checks */
+   if (W_offd_nnz > 0 || W_offd_num_cols > 0)
+   {
+      *R_ptr = NULL;
+
+      hypre_error_w_msg(HYPRE_ERROR_GENERIC, "Not implemented for matrices with nonzero offd");
+      return hypre_error_flag;
+   }
+
+   /*-----------------------------------------------------------------------
+    * Create and initialize output matrix
+    *-----------------------------------------------------------------------*/
+
+   R = hypre_ParCSRMatrixCreate(comm, num_rows_R, num_cols_R,
+                                row_starts_R, col_starts_R, W_offd_num_cols,
+                                num_nonzeros_diag, num_nonzeros_offd);
+   hypre_ParCSRMatrixInitialize_v2(R, memory_location_W);
+
+   /*-----------------------------------------------------------------------
+    * Fill matrix entries
+    *-----------------------------------------------------------------------*/
+
+#if defined (HYPRE_USING_GPU)
+   HYPRE_ExecutionPolicy exec = hypre_GetExecPolicy1(memory_location_W);
+
+   if (exec == HYPRE_EXEC_DEVICE)
+   {
+      /* TODO (VPM): Implement hypre_MGRBuildRFromWDevice */
+      hypre_ParCSRMatrixMigrate(W, HYPRE_MEMORY_HOST);
+      hypre_ParCSRMatrixMigrate(R, HYPRE_MEMORY_HOST);
+      hypre_MGRBuildRFromWHost(C_map, F_map, W, R);
```

Comment:
Do you think we need a warning here? Not sure what we do in similar cases there there's no device option yet and we revert to running on host?

Reply:
I think it isn't necessary as the code still works despite being less efficient than running fully on the device. I'm planning to implement `hypre_MGRBuildRFromWDevice` in the next couple of months, so this is temporary

Reply:
Ok sounds good.